File: AmericanExpress/src/main/java/com/braintreepayments/api/americanexpress/AmericanExpressClient.java
package com.braintreepayments.api.americanexpress;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.annotation.VisibleForTesting;

import com.braintreepayments.api.ApiClient;
import com.braintreepayments.api.core.BraintreeClient;

import org.json.JSONException;

/**
 * Used to integrate with Braintree's American Express API
 */
public class AmericanExpressClient {

    private static final String AMEX_REWARDS_BALANCE_PATH =
            ApiClient.versionedPath("payment_methods/amex_rewards_balance");

    private final BraintreeClient braintreeClient;

    /**
     * Initializes a new {@link AmericanExpressClient} instance
     *
     * @param context an Android Context
     * @param authorization a Tokenization Key or Client Token used to authenticate
     */
    public AmericanExpressClient(@NonNull Context context, @NonNull String authorization) {
        this.braintreeClient = new BraintreeClient(context, authorization);
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public AmericanExpressClient(@NonNull BraintreeClient braintreeClient) {
        this.braintreeClient = braintreeClient;
    }

    /**
     * Gets the rewards balance associated with a Braintree nonce. Only for American Express cards.
     *
     * @param nonce           A nonce representing a card that will be used to look up the rewards
     *                        balance
     * @param currencyIsoCode The currencyIsoCode to use. Example: 'USD'
     * @param callback        {@link AmericanExpressGetRewardsBalanceCallback}
     */
    public void getRewardsBalance(@NonNull String nonce, @NonNull String currencyIsoCode, @NonNull
    final AmericanExpressGetRewardsBalanceCallback callback) {
        String getRewardsBalanceUrl = Uri.parse(AMEX_REWARDS_BALANCE_PATH)
                .buildUpon()
                .appendQueryParameter("paymentMethodNonce", nonce)
                .appendQueryParameter("currencyIsoCode", currencyIsoCode)
                .build()
                .toString();

        braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED);
        braintreeClient.sendGET(getRewardsBalanceUrl, (responseBody, httpError) -> {
            if (responseBody != null) {
                try {
                    AmericanExpressRewardsBalance rewardsBalance =
                            AmericanExpressRewardsBalance.fromJson(responseBody);
                    callbackSuccess(new AmericanExpressResult.Success(rewardsBalance), callback);
                } catch (JSONException e) {
                    callbackFailure(new AmericanExpressResult.Failure(e), callback);
                }
            } else if (httpError != null) {
                callbackFailure(new AmericanExpressResult.Failure(httpError), callback);
            }
        });
    }

    private void callbackSuccess(AmericanExpressResult.Success result, AmericanExpressGetRewardsBalanceCallback callback) {
        callback.onAmericanExpressResult(result);
        braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_SUCCEEDED);
    }

    private void callbackFailure(AmericanExpressResult.Failure result, AmericanExpressGetRewardsBalanceCallback callback) {
        callback.onAmericanExpressResult(result);
        braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_FAILED);
    }
}

File: AmericanExpress/src/main/java/com/braintreepayments/api/americanexpress/AmericanExpressClient.kt
package com.braintreepayments.api.americanexpress

import android.content.Context
import android.net.Uri
import com.braintreepayments.api.core.AnalyticsEventParams
import com.braintreepayments.api.core.ApiClient.Companion.versionedPath
import com.braintreepayments.api.core.BraintreeClient
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Used to integrate with Braintree's American Express API
 */
class AmericanExpressClient internal constructor(
    private val braintreeClient: BraintreeClient,
    private val mainDispatcher: CoroutineDispatcher = Dispatchers.Main,
    private val coroutineScope: CoroutineScope = CoroutineScope(mainDispatcher),
) {

    /**
     * Initializes a new [AmericanExpressClient] instance
     *
     * @param context an Android Context
     * @param authorization a Tokenization Key or Client Token used to authenticate
     */
    constructor(context: Context, authorization: String) : this(
        BraintreeClient(
            context,
            authorization
        )
    )

    /**
     * Gets the rewards balance associated with a Braintree nonce. Only for American Express cards.
     *
     * @param nonce   A nonce representing a card that will be used to look up the rewards balance
     * @param currencyIsoCode The currencyIsoCode to use. Example: 'USD'
     * @param callback        [AmericanExpressGetRewardsBalanceCallback]
     */
    fun getRewardsBalance(
        nonce: String,
        currencyIsoCode: String,
        callback: AmericanExpressGetRewardsBalanceCallback
    ) {
        coroutineScope.launch {
            val result = getRewardsBalance(nonce, currencyIsoCode)
            callback.onAmericanExpressResult(result)
        }
    }

    private suspend fun getRewardsBalance(
        nonce: String,
        currencyIsoCode: String
    ): AmericanExpressResult {
        val getRewardsBalanceUrl = Uri.parse(AMEX_REWARDS_BALANCE_PATH)
            .buildUpon()
            .appendQueryParameter("paymentMethodNonce", nonce)
            .appendQueryParameter("currencyIsoCode", currencyIsoCode)
            .build()
            .toString()

        braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_STARTED)
        try {
            val responseBody = braintreeClient.sendGET(getRewardsBalanceUrl)
            val rewardsBalance =
                AmericanExpressRewardsBalance.fromJson(responseBody)
            braintreeClient.sendAnalyticsEvent(AmericanExpressAnalytics.REWARDS_BALANCE_SUCCEEDED)
            return AmericanExpressResult.Success(rewardsBalance)
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            return getRewardsBalanceFailure(AmericanExpressResult.Failure(e))
        }
    }

    private fun getRewardsBalanceFailure(
        result: AmericanExpressResult.Failure,
    ): AmericanExpressResult {
        braintreeClient.sendAnalyticsEvent(
            AmericanExpressAnalytics.REWARDS_BALANCE_FAILED,
            AnalyticsEventParams(errorDescription = result.error.message)
        )
        return result
    }

    companion object {
        private val AMEX_REWARDS_BALANCE_PATH =
            versionedPath("payment_methods/amex_rewards_balance")
    }
}
